File: astronomy.py
"""
astronomy.py — Core Swiss Ephemeris wrapper for the Panchang Generator.

Provides:
  - Julian Date conversion
  - Ayanamsa retrieval (Lahiri, Raman, Krishnamurti)
  - Nirayana (sidereal) planet longitudes with DMS formatting
  - Sunrise / Sunset / Moonrise / Moonset calculation
  - Julian Day → local time string conversion
"""

import swisseph as swe
import math

# ---------------------------------------------------------------------------
# Setup
# ---------------------------------------------------------------------------
# Point to ephemeris files if available; empty string → built-in Moshier
swe.set_ephe_path('')

# ---------------------------------------------------------------------------
# Constants & Mappings
# ---------------------------------------------------------------------------
AYANAMSA_SYSTEMS: dict[str, int] = {
    'Lahiri':       swe.SIDM_LAHIRI,
    'Raman':        swe.SIDM_RAMAN,
    'Krishnamurti': swe.SIDM_KRISHNAMURTI,
}

PLANETS: dict[str, int] = {
    'Sun':     swe.SUN,
    'Moon':    swe.MOON,
    'Mars':    swe.MARS,
    'Mercury': swe.MERCURY,
    'Jupiter': swe.JUPITER,
    'Venus':   swe.VENUS,
    'Saturn':  swe.SATURN,
    'Rahu':    swe.TRUE_NODE,   # True Lunar North Node (more accurate)
    # Ketu is computed as Rahu + 180°
}

# ---------------------------------------------------------------------------
# Julian Date helpers
# ---------------------------------------------------------------------------

def get_julian_date(year: int, month: int, day: int, hour_utc: float) -> float:
    """Return the Julian Day Number for a given UTC datetime.

    Args:
        year, month, day: Calendar date.
        hour_utc: Time in UTC as a decimal fraction (e.g., 5.5 = 05:30 UTC).
    """
    return swe.julday(year, month, day, hour_utc)


def local_time_to_jd(year: int, month: int, day: int,
                     local_hour: float, tz_offset: float = 5.5) -> float:
    """Convert a local time to Julian Day (UTC-based).

    Args:
        local_hour:  Local clock time as a decimal (e.g., 5.5 = 05:30).
        tz_offset:   Hours east of UTC (default 5.5 = IST).
    """
    hour_utc = local_hour - tz_offset
    # Handle crossing into the previous or next UTC day
    delta_days = 0
    if hour_utc < 0:
        hour_utc += 24.0
        delta_days = -1
    elif hour_utc >= 24.0:
        hour_utc -= 24.0
        delta_days = 1
    jd = swe.julday(year, month, day, hour_utc)
    return jd + delta_days


# ---------------------------------------------------------------------------
# Ayanamsa
# ---------------------------------------------------------------------------

def get_ayanamsa(julian_date: float, ayanamsa_name: str = 'Lahiri') -> float:
    """Return the ayanamsa value in decimal degrees for the given JD."""
    sid_mode = AYANAMSA_SYSTEMS.get(ayanamsa_name, swe.SIDM_LAHIRI)
    swe.set_sid_mode(sid_mode, 0, 0)
    return swe.get_ayanamsa_ut(julian_date)


# ---------------------------------------------------------------------------
# Planetary Positions
# ---------------------------------------------------------------------------

def _set_sidereal_mode(ayanamsa_name: str) -> None:
    """Internal helper: set sidereal mode before a calculation."""
    sid_mode = AYANAMSA_SYSTEMS.get(ayanamsa_name, swe.SIDM_LAHIRI)
    swe.set_sid_mode(sid_mode, 0, 0)


def get_planetary_longitude(julian_date: float,
                             planet_name: str,
                             ayanamsa_name: str = 'Lahiri') -> float:
    """Return the Nirayana (sidereal) longitude of a planet in decimal degrees.

    Uses geocentric apparent positions with high-precision Swiss Ephemeris mode.

    Args:
        julian_date:   Julian Day Number (UT).
        planet_name:   One of 'Sun', 'Moon', 'Mars', 'Mercury', 'Jupiter',
                       'Venus', 'Saturn', 'Rahu', 'Ketu'.
        ayanamsa_name: Ayanamsa system to use.

    Returns:
        Longitude in decimal degrees [0, 360).
    """
    if planet_name == 'Ketu':
        rahu_lon = get_planetary_longitude(julian_date, 'Rahu', ayanamsa_name)
        return (rahu_lon + 180.0) % 360.0

    planet_id = PLANETS[planet_name]
    _set_sidereal_mode(ayanamsa_name)

    flags = swe.FLG_SWIEPH | swe.FLG_SIDEREAL | swe.FLG_SPEED
    res, _err = swe.calc_ut(julian_date, planet_id, flags)
    return res[0] % 360.0


def get_all_planet_positions(julian_date: float,
                              ayanamsa_name: str = 'Lahiri') -> dict[str, float]:
    """Return sidereal longitudes for all 9 Vedic grahas.

    Returns:
        Dict mapping planet name → decimal degrees.
    """
    _set_sidereal_mode(ayanamsa_name)
    flags = swe.FLG_SWIEPH | swe.FLG_SIDEREAL | swe.FLG_SPEED

    result: dict[str, float] = {}
    for name, pid in PLANETS.items():
        res, _err = swe.calc_ut(julian_date, pid, flags)
        result[name] = res[0] % 360.0

    # Ketu = Rahu + 180°
    result['Ketu'] = (result['Rahu'] + 180.0) % 360.0
    return result


# ---------------------------------------------------------------------------
# DMS formatting
# ---------------------------------------------------------------------------

def decimal_to_dms(decimal_degrees: float) -> tuple[int, int, float]:
    """Convert decimal degrees → (degrees, minutes, seconds)."""
    d = int(decimal_degrees)
    m_dec = (decimal_degrees - d) * 60.0
    m = int(m_dec)
    s = (m_dec - m) * 60.0
    return d, m, s


def format_dms(decimal_degrees: float) -> str:
    """Return a human-readable DMS string, e.g. '123° 45′ 06.78″'."""
    d, m, s = decimal_to_dms(decimal_degrees)
    return f"{d:3d}° {m:02d}' {s:05.2f}\""


def get_rashi_name(decimal_degrees: float) -> str:
    """Return the Rashi (zodiac sign) name for a given sidereal longitude."""
    RASHI_NAMES = [
        "Mesha (Aries)", "Vrishabha (Taurus)", "Mithuna (Gemini)",
        "Karka (Cancer)", "Simha (Leo)", "Kanya (Virgo)",
        "Tula (Libra)", "Vrishchika (Scorpio)", "Dhanu (Sagittarius)",
        "Makara (Capricorn)", "Kumbha (Aquarius)", "Meena (Pisces)",
    ]
    idx = int(decimal_degrees / 30.0) % 12
    return RASHI_NAMES[idx]


# ---------------------------------------------------------------------------
# Rise / Set calculations
# ---------------------------------------------------------------------------

def _rise_set(julian_date: float, body: int,
              lat: float, lon: float,
              is_rise: bool) -> float:
    """Internal: compute rise or set JD for a celestial body.

    Args:
        julian_date: JD used as search start.
        body:        Swiss Ephemeris body constant.
        lat, lon:    Geographic coordinates (degrees, positive north/east).
        is_rise:     True → rise time; False → set time.

    Returns:
        Julian Day of the event, or 0.0 on failure.
    """
    geopos = (lon, lat, 0.0)       # (longitude, latitude, altitude_m)

    # Start from 00:00 UTC of the given calendar day
    # swe.julday gives JD at 12:00 UTC (noon), subtract 0.5 to get midnight
    y, m, d, _ = swe.revjul(julian_date, swe.GREG_CAL)
    jd_start = swe.julday(y, m, d, 0.0)   # 00:00 UTC of that date

    rsmi = swe.CALC_RISE if is_rise else swe.CALC_SET
    rsmi |= swe.BIT_DISC_CENTER

    try:
        # pyswisseph 2.10.x signature:
        # rise_trans(tjdut, body, rsmi, geopos, atpress=0, attemp=0, flags=FLG_SWIEPH)
        # Note: no star_name argument in this version
        ret, tret = swe.rise_trans(
            jd_start,
            body,
            rsmi,
            geopos,
            1013.25,
            15.0,
        )
        if ret == 0 and tret[0] > 0:
            return tret[0]
    except Exception:
        pass

    # Fallback without disc-center correction
    try:
        rsmi2 = swe.CALC_RISE if is_rise else swe.CALC_SET
        ret, tret = swe.rise_trans(jd_start, body, rsmi2, geopos, 1013.25, 15.0)
        if ret == 0 and tret[0] > 0:
            return tret[0]
    except Exception:
        pass

    return 0.0


def get_sunrise(julian_date: float, lat: float, lon: float) -> float:
    """Return Julian Day of sunrise for the given date and location."""
    return _rise_set(julian_date, swe.SUN, lat, lon, is_rise=True)


def get_sunset(julian_date: float, lat: float, lon: float) -> float:
    """Return Julian Day of sunset for the given date and location."""
    return _rise_set(julian_date, swe.SUN, lat, lon, is_rise=False)


def get_moonrise(julian_date: float, lat: float, lon: float) -> float:
    """Return Julian Day of moonrise for the given date and location."""
    return _rise_set(julian_date, swe.MOON, lat, lon, is_rise=True)


def get_moonset(julian_date: float, lat: float, lon: float) -> float:
    """Return Julian Day of moonset for the given date and location."""
    return _rise_set(julian_date, swe.MOON, lat, lon, is_rise=False)


# ---------------------------------------------------------------------------
# Time Conversion
# ---------------------------------------------------------------------------

def jd_to_local_time_string(jd: float, tz_offset: float = 5.5) -> str:
    """Convert a Julian Day to a local time string HH:MM:SS.

    Args:
        jd:         Julian Day Number.
        tz_offset:  Hours east of UTC (default 5.5 = IST UTC+5:30).

    Returns:
        Local time string like '07:18:34', or '--:--:--' on failure.
    """
    if jd == 0.0:
        return '--:--:--'
    try:
        _y, _m, _d, h_utc = swe.revjul(jd, swe.GREG_CAL)
        h_local = h_utc + tz_offset
        # Handle overflow across midnight
        h_local = h_local % 24.0
        hh = int(h_local)
        mm_float = (h_local - hh) * 60.0
        mm = int(mm_float)
        ss = int((mm_float - mm) * 60.0)
        return f"{hh:02d}:{mm:02d}:{ss:02d}"
    except Exception:
        return '--:--:--'


# Backward-compatible alias used by old callers
def jd_to_ist_string(jd: float) -> str:
    """Convert a Julian Day to IST (UTC+5:30) time string."""
    return jd_to_local_time_string(jd, tz_offset=5.5)

File: android_app/app/src/main/java/com/example/jainpanchang/domain/Astronomy.kt
package com.example.jainpanchang.domain

import de.thmac.swisseph.SweConst
import de.thmac.swisseph.SwissEph
import de.thmac.swisseph.SweDate
import de.thmac.swisseph.DblObj
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZonedDateTime

class Astronomy {
    private val swe = SwissEph()

    init {
        // Points to internal ephemeris Moshier by default
        swe.swe_set_ephe_path("")
    }

    val AYANAMSA_SYSTEMS = mapOf(
        "Lahiri" to SweConst.SE_SIDM_LAHIRI,
        "Raman" to SweConst.SE_SIDM_RAMAN,
        "Krishnamurti" to SweConst.SE_SIDM_KRISHNAMURTI
    )

    val RASHI_NAMES = listOf(
        "Mesha (Aries)", "Vrishabha (Taurus)", "Mithuna (Gemini)",
        "Karka (Cancer)", "Simha (Leo)", "Kanya (Virgo)",
        "Tula (Libra)", "Vrishchika (Scorpio)", "Dhanu (Sagittarius)",
        "Makara (Capricorn)", "Kumbha (Aquarius)", "Meena (Pisces)"
    )

    val SUN_RASHI_NAMES = listOf(
        "Mesh", "Vrishabh", "Mithun", "Kark", "Simha", "Kanya",
        "Tula", "Vrishchik", "Dhanu", "Makar", "Kumbh", "Meen"
    )

    val PLANETS = mapOf(
        "Sun" to SweConst.SE_SUN,
        "Moon" to SweConst.SE_MOON,
        "Mars" to SweConst.SE_MARS,
        "Mercury" to SweConst.SE_MERCURY,
        "Jupiter" to SweConst.SE_JUPITER,
        "Venus" to SweConst.SE_VENUS,
        "Saturn" to SweConst.SE_SATURN,
        "Rahu" to SweConst.SE_TRUE_NODE
    )

    fun getJulianDate(year: Int, month: Int, day: Int, hourUtc: Double): Double {
        val sd = SweDate(year, month, day, hourUtc, true) // true for Gregorian
        return sd.getJulDay()
    }

    fun localTimeToJd(year: Int, month: Int, day: Int, localHour: Double, tzOffset: Double = 5.5): Double {
        var hourUtc = localHour - tzOffset
        var deltaDays = 0
        if (hourUtc < 0) {
            hourUtc += 24.0
            deltaDays = -1
        } else if (hourUtc >= 24.0) {
            hourUtc -= 24.0
            deltaDays = 1
        }
        val sd = SweDate(year, month, day, hourUtc, true)
        return sd.getJulDay() + deltaDays
    }

    fun zonedDatetimeToJd(localDt: ZonedDateTime): Double {
        val utcDt = localDt.withZoneSameInstant(ZoneId.of("UTC"))
        val hourUtc = utcDt.hour + utcDt.minute / 60.0 + utcDt.second / 3600.0 + utcDt.nano / 3_600_000_000_000.0
        val sd = SweDate(utcDt.year, utcDt.monthValue, utcDt.dayOfMonth, hourUtc, true)
        return sd.getJulDay()
    }

    fun localDateAnchorJd(localDate: LocalDate, tzName: String, hour: Int = 0): Double {
        val localDt = localDate.atTime(hour, 0).atZone(ZoneId.of(tzName))
        return zonedDatetimeToJd(localDt)
    }

    fun getAyanamsa(julianDate: Double, ayanamsaName: String = "Lahiri"): Double {
        val sidMode = AYANAMSA_SYSTEMS[ayanamsaName] ?: SweConst.SE_SIDM_LAHIRI
        swe.swe_set_sid_mode(sidMode, 0.0, 0.0)
        return swe.swe_get_ayanamsa_ut(julianDate)
    }

    private fun setSiderealMode(ayanamsaName: String) {
        val sidMode = AYANAMSA_SYSTEMS[ayanamsaName] ?: SweConst.SE_SIDM_LAHIRI
        swe.swe_set_sid_mode(sidMode, 0.0, 0.0)
    }

    fun getPlanetaryLongitude(julianDate: Double, planetName: String, ayanamsaName: String = "Lahiri"): Double {
        if (planetName == "Ketu") {
            val rahuLon = getPlanetaryLongitude(julianDate, "Rahu", ayanamsaName)
            return (rahuLon + 180.0) % 360.0
        }
        val planetId = PLANETS[planetName] ?: return 0.0
        setSiderealMode(ayanamsaName)
        
        val flags = SweConst.SEFLG_SWIEPH or SweConst.SEFLG_SIDEREAL or SweConst.SEFLG_SPEED
        val res = DoubleArray(6)
        val err = StringBuffer()
        swe.swe_calc_ut(julianDate, planetId, flags, res, err)
        return res[0] % 360.0
    }

    fun getAllPlanetPositions(julianDate: Double, ayanamsaName: String = "Lahiri"): Map<String, Double> {
        val result = mutableMapOf<String, Double>()
        setSiderealMode(ayanamsaName)
        val flags = SweConst.SEFLG_SWIEPH or SweConst.SEFLG_SIDEREAL or SweConst.SEFLG_SPEED

        for ((name, pid) in PLANETS) {
            val res = DoubleArray(6)
            val err = StringBuffer()
            swe.swe_calc_ut(julianDate, pid, flags, res, err)
            result[name] = res[0] % 360.0
        }
        result["Ketu"] = ((result["Rahu"] ?: 0.0) + 180.0) % 360.0
        return result
    }

    fun decimalToDms(decimalDegrees: Double): Triple<Int, Int, Double> {
        val d = decimalDegrees.toInt()
        val mDec = (decimalDegrees - d) * 60.0
        val m = mDec.toInt()
        val s = (mDec - m) * 60.0
        return Triple(d, m, s)
    }

    fun formatDms(decimalDegrees: Double): String {
        val (d, m, s) = decimalToDms(decimalDegrees)
        return String.format("%3d° %02d' %05.2f\"", d, m, s)
    }

    fun getRashiName(decimalDegrees: Double): String {
        val idx = (decimalDegrees / 30.0).toInt() % 12
        return RASHI_NAMES[idx]
    }

    fun getSunRashi(jdSunrise: Double): String {
        val flags = SweConst.SEFLG_SWIEPH or SweConst.SEFLG_SPEED
        val res = DoubleArray(6)
        val err = StringBuffer()
        swe.swe_calc_ut(jdSunrise, SweConst.SE_SUN, flags, res, err)
        val tropicalLon = res[0]
        swe.swe_set_sid_mode(SweConst.SE_SIDM_LAHIRI, 0.0, 0.0)
        val ayanamsa = swe.swe_get_ayanamsa_ut(jdSunrise)
        val siderealLon = (tropicalLon - ayanamsa + 360.0) % 360.0
        val idx = (siderealLon / 30.0).toInt() % 12
        return SUN_RASHI_NAMES[idx]
    }

    private fun riseSet(julianDate: Double, body: Int, lat: Double, lon: Double, isRise: Boolean): Double {
        val geopos = doubleArrayOf(lon, lat, 0.0)
        val rsmi = if (isRise) SweConst.SE_CALC_RISE else SweConst.SE_CALC_SET
        
        val tret = DblObj()
        val err = StringBuffer()
        
        try {
            val ret = swe.swe_rise_trans(
                julianDate,
                body,
                StringBuffer(""),
                SweConst.SEFLG_SWIEPH,
                rsmi,
                geopos,
                1013.25,
                15.0,
                tret,
                err
            )
            if (ret == 0 && tret.`val` > 0) {
                return tret.`val`
            }
        } catch (e: Exception) {}
        
        return 0.0
    }

    fun getSunrise(julianDate: Double, lat: Double, lon: Double): Double {
        return riseSet(julianDate, SweConst.SE_SUN, lat, lon, true)
    }

    fun getSunset(julianDate: Double, lat: Double, lon: Double): Double {
        return riseSet(julianDate, SweConst.SE_SUN, lat, lon, false)
    }

    fun getMoonrise(julianDate: Double, lat: Double, lon: Double): Double {
        return riseSet(julianDate, SweConst.SE_MOON, lat, lon, true)
    }

    fun getMoonset(julianDate: Double, lat: Double, lon: Double): Double {
        return riseSet(julianDate, SweConst.SE_MOON, lat, lon, false)
    }

    fun jdToLocalTimeString(jd: Double, tzOffset: Double = 5.5): String {
        if (jd == 0.0) return "--:--:--"
        try {
            val s = SweDate(jd)
            var hLocal = s.hour + tzOffset
            hLocal %= 24.0
            if (hLocal < 0) hLocal += 24.0
            val hh = hLocal.toInt()
            val mmFloat = (hLocal - hh) * 60.0
            val mm = mmFloat.toInt()
            val ss = ((mmFloat - mm) * 60.0).toInt()
            return String.format("%02d:%02d:%02d", hh, mm, ss)
        } catch (e: Exception) {
            return "--:--:--"
        }
    }

    fun buildEclipseDateSets(startJd: Double, endJd: Double, lat: Double, lon: Double): Pair<Set<String>, Set<String>> {
        val geopos = doubleArrayOf(lon, lat, 0.0)
        val solarDates = mutableSetOf<String>()
        val lunarDates = mutableSetOf<String>()

        var jd = startJd
        val tret = DoubleArray(10)
        val attr = DoubleArray(20)
        val err = StringBuffer()

        while (jd < endJd) {
            try {
                val ret = swe.swe_sol_eclipse_when_loc(jd, SweConst.SEFLG_SWIEPH, geopos, tret, attr, 0, err)
                if (ret > 0 && tret[0] > 0) {
                    val s = SweDate(tret[0])
                    solarDates.add(String.format("%04d-%02d-%02d", s.year, s.month, s.day))
                    jd = tret[0] + 20.0
                } else {
                    jd += 150.0
                }
            } catch (e: Exception) {
                jd += 150.0
            }
        }

        jd = startJd
        while (jd < endJd) {
            try {
                val ret = swe.swe_lun_eclipse_when_loc(jd, SweConst.SEFLG_SWIEPH, geopos, tret, attr, 0, err)
                if (ret > 0 && tret[0] > 0) {
                    val s = SweDate(tret[0])
                    lunarDates.add(String.format("%04d-%02d-%02d", s.year, s.month, s.day))
                    jd = tret[0] + 20.0
                } else {
                    jd += 25.0
                }
            } catch (e: Exception) {
                jd += 25.0
            }
        }

        return Pair(solarDates, lunarDates)
    }
}
